allennlp/modules/transformer/t5.py

```diff
@@ -773,187 +801,473 @@ def forward(
-            # get decoder inputs from shifting lm labels to the right
-            decoder_input_ids = self._shift_right(labels)
-
-        # If decoding with past key value states, only the last tokens
-        # should be given as an input
-        if past_key_values is not None:
-            assert labels is None, "Decoder should not use cached key value states when training."
-            if decoder_input_ids is not None:
-                decoder_input_ids = decoder_input_ids[:, -1:]
-            if decoder_inputs_embeds is not None:
-                decoder_inputs_embeds = decoder_inputs_embeds[:, -1:]
-
-        # Decode
-        decoder_outputs: T5StackOutput = self.decoder(
-            input_ids=decoder_input_ids,
-            attention_mask=decoder_attention_mask,
-            inputs_embeds=decoder_inputs_embeds,
-            past_key_values=past_key_values,
-            encoder_hidden_states=hidden_states,
-            encoder_attention_mask=attention_mask,
-            head_mask=decoder_head_mask,
-            encoder_head_mask=head_mask,
-            use_cache=use_cache,
-            output_attentions=output_attentions,
-            output_hidden_states=output_hidden_states,
+        input_ids: IntT,
+        attention_mask: Optional[BoolT] = None,
+        labels: Optional[IntT] = None,
+        decoder_attention_mask: Optional[BoolT] = None,
+    ) -> T5Output:
+        """
+        Run forward pass of the model.
+        """
+        if attention_mask is None:
+            attention_mask = ~(input_ids == self.pad_token_id)
+
+        # Encode inputs.
+        encoder_outputs: T5StackOutput = self.encoder(
+            input_ids=input_ids,
+            attention_mask=attention_mask,
+            output_attentions=self.output_attentions,
+            output_all_hidden_states=self.output_all_hidden_states,
         )
 
-        lm_logits = self.lm_head(decoder_outputs.hidden_states)
+        logits: Optional[FloatT] = None
+        loss: Optional[FloatT] = None
+        decoder_outputs: Optional[T5StackOutput] = None
+        predictions: Optional[IntT] = None
+        predicted_log_probs: Optional[FloatT] = None
 
-        loss = None
         if labels is not None:
-            loss_fct = CrossEntropyLoss(ignore_index=-100)
-            loss = loss_fct(lm_logits.view(-1, lm_logits.size(-1)), labels.view(-1))
+            # Calculate loss against targets.
+
+            if decoder_attention_mask is None:
+                decoder_attention_mask = ~(labels == self.pad_token_id)
+
+            # Get decoder inputs from shifting lm labels to the right and pre-pending
+            # the decoder start token ID.
+            # Shape (both): (batch_size, target_length)
+            decoder_input_ids = self._shift_right(labels, self.decoder_start_token_id)
+
+            # Replace possible -100 values in labels by `pad_token_id`
+            decoder_input_ids.masked_fill_(decoder_input_ids == -100, self.pad_token_id)
```

Comment:
A backwards compatibility hack we carried over I assume?

Reply:
Carried over, but I don't know why 😏